### ISSUE TITLE
Bazel EP integration: Warnings from the wrong projects are reported

### ISSUE THREAD

**Issue by davido** [google/error-prone#664]
Gerrit Code Review project is using Bazel as build tool. Since recently gerrit depends on `rules_closure` for JS tool chain.

Given that Bazel's EP integration is missing the ability to activate **all** warnings, so that this patch is needed: [1],[2]. After applying that patch and building gerrit with patched Bazel version, I'm getting hundred of warnings on `rules_closure` and Bazel sources.

Needless to say, that I do not care about any EP issues in any other projects that gerrit itself. How can I suppress warnings in all *foreign* projects?

FYI: I reported this issue upstream: https://github.com/bazelbuild/rules_closure/issues/211, where I described the steps to reproduce the problem.

* [1] https://github.com/bazelbuild/bazel/issues/2237 
* [2] https://bazel-review.googlesource.com/7891

**Comment by cushon** [comment]
Are you using [`--output_filter`](https://docs.bazel.build/versions/master/bazel-user-manual.html#flag--output_filter)?

This is part of the reason they're disabled by default - they tend to make build logs very noisy.

**Comment by davido** [comment]
I'm not using `--output_filter`. Before depending on `rules_closure`, with [1] applied, and without any further option (`bazel build gerrit`) we were getting reasonable amount of warnings in gerrit code only. After adding dependency we are getting spam on `rules_closure` sources and Bazel code itself, that is also with any additional options, just with `bazel build gerrit`.

* [1] https://bazel-review.googlesource.com/7891

**Comment by cushon** [comment]
To confirm, the spam you're referring to is additional build output for the warnings that were enabled? This is the problem that `--output_filter` is supposed to prevent, but the default behaviour for Bazel isn't what I expected. I filed bazelbuild/bazel#3330 for that. Do you have suggestions for changes in Error Prone itself that would help here?

**Comment by davido** [comment]
> To confirm, the spam you're referring to is additional build output for the warnings that were enabled? 

This is correct. Unless I activate all warnings per default with my Bazel patch: [1], I don't get spammed from `rules_closure` and Bazel itself. In fact without my patch I do not get any EP warnings at all, as I'm not going to activate single EP warnings for the reasons explained in: [1]. However, just guessing If I would try to activate warning per WP options, then I would see the same spam from the wrong projects (`rules_closure` in gerrit case). As I said, I haven't verified that.

> Do you have suggestions for changes in Error Prone itself that would help here?

Sure. I would go with the approach from Facebook's Infer tool: http://fbinfer.com/docs/infer-workflow.html and would try to borrow their path-black-listing configuration approach, like:

https://github.com/facebook/buck/blob/master/.inferconfig#L6-L8

To apply this approach to gerri's rules_closure all default warning spam, i would say something like in .errorprone:

```
"errorprone-blacklist-path-regex": [
   "external/io_bazel_rules_closure/.*"
],
```

That way, the notification firehouse from foreign projects would be turned off, even with all default EP warnings activated:

```
NFO: From Building external/io_bazel_rules_closure/java/io/bazel/rules/closure/webfiles/libbuild_info_java_proto.jar (1 source jar) [for host]:
/io/bazel/rules/closure/webfiles/BuildInfo.java:775: warning: [FunctionalInterfaceClash] When passing lambda arguments to this function, callers will need a cast to disambiguate with: GeneratedMessageV3.newBuilderForType(BuilderParent)
    protected Builder newBuilderForType(
                      ^
      AbstractMessage.newBuilderForType(BuilderParent)
    (see http://errorprone.info/bugpattern/FunctionalInterfaceClash)
```

* [1] https://bazel-review.googlesource.com/7891

**Comment by cushon** [comment]
> However, just guessing If I would try to activate warning per EP options, then I would see the same spam from the wrong projects...

That's what I'd expect, both ways of enabling them are equivalent. Both cause Bazel to enable them for all compilations.

> I would go with the approach from Facebook's Infer tool

Thanks for the pointer. That seems similar to `--output_filter` in that you can use regexes to control which packages output is emitted for. It offers more control in suppressing output for specific checks. Are there other differences?

**Comment by cushon** [comment]
Also - there's some related discussion in google/error-prone#648, in particular https://github.com/google/error-prone/issues/648#issuecomment-310761743

**Comment by davido** [comment]
> Are there other differences?

The configuration with `.errorprone` specific configuration file would not be Bazel specific but would work for all EP integrations.

It would be less awkward to instruct the contributors (and maintain this information). We would say in our contribution documentation with all default EP warnings: `bazel build --javacopts=EP-ALL-WARN-ACTIVE=true gerrit`, done. And we (maintainer) would take care our self to maintain the right `.errorprone` and blacklist the right packages (say we add other dependencies other than `rules_closure` that spam the output).

**Comment by cushon** [comment]
Note that for Bazel you can already check in a `.bazelrc` that will be shared with anyone who uses the project.

**Comment by davido** [comment]
Ah, right, thanks for pointing this out. In gerrt we even have the config already `[gerrit-root-directory]/tools/bazel.rc`: [1], so that we would just extend it.

[1] https://github.com/GerritCodeReview/gerrit/blob/master/tools/bazel.rc#L1
